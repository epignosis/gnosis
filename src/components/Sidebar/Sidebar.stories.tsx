import React, { ReactNode } from "react";
import { Story } from "@storybook/react";
import { Avatar } from "../../";
import SidebarComponent, { SidebarProps } from "./Sidebar";
import {
  CalendarSVG,
  CatalogSVG,
  ChatCloudsSVG,
  CommunicationSVG,
  HelpSVG,
  HomeSVG,
  MyCoursesSVG,
  OpenedMessageSVG,
} from "@icons/core";

export default {
  component: SidebarComponent,
  title: "Components/Sidebar",
  decorators: [
    (story: () => ReactNode): JSX.Element => <div style={{ display: "flex" }}>{story()}</div>,
  ],
  argTypes: {
    onToggle: {
      action: "Toggle action fired",
    },
  },
  args: {
    isCollapsed: false,
    style: { zIndex: 10, overflow: "auto" },
  },
};

export const Sidebar: Story<SidebarProps> = (args) => (
  <SidebarComponent {...args}>
    <div
      style={{ display: "flex", flexDirection: "column", justifyContent: "space-between", flex: 1 }}
    >
      <ul style={{ listStyle: "none", padding: 0, margin: 0 }}>
        <li>
          <a href="www.google.com" style={{ display: "block" }}>
            <SidebarComponent.Item
              icon={<HomeSVG height={32} />}
              label="Home"
              isExpanded={!args.isCollapsed}
            />
          </a>
        </li>
        <li>
          <a href="www.google.com" style={{ display: "block" }}>
            <SidebarComponent.Item
              icon={<MyCoursesSVG height={32} />}
              label="My courses"
              isExpanded={!args.isCollapsed}
            />
          </a>
        </li>
        <li>
          <a href="www.google.com" style={{ display: "block" }}>
            <SidebarComponent.Item
              icon={<CatalogSVG height={32} />}
              label="Catalog"
              isExpanded={!args.isCollapsed}
            />
          </a>
        </li>
        <li>
          <a href="www.google.com" style={{ display: "block" }}>
            <SidebarComponent.Item
              icon={<CalendarSVG height={32} />}
              label="Calendar"
              isExpanded={!args.isCollapsed}
            />
          </a>
        </li>
      </ul>

      <ul style={{ listStyle: "none", padding: 0, margin: 0 }}>
        <li>
          <a href="www.google.com" style={{ display: "block" }}>
            <SidebarComponent.Item
              icon={<OpenedMessageSVG height={32} />}
              label="Messages"
              isExpanded={!args.isCollapsed}
            />
          </a>
        </li>
        <li>
          <a href="www.google.com" style={{ display: "block" }}>
            <SidebarComponent.Item
              icon={<ChatCloudsSVG height={32} />}
              label="Discussions"
              isExpanded={!args.isCollapsed}
            />
          </a>
        </li>
        <li>
          <a href="www.google.com" style={{ display: "block" }}>
            <SidebarComponent.Item
              icon={<CommunicationSVG height={32} />}
              label="Notifications"
              isExpanded={!args.isCollapsed}
            />
          </a>
        </li>
        <li>
          <a href="www.google.com" style={{ display: "block" }}>
            <SidebarComponent.Item
              icon={<HelpSVG height={32} />}
              label="Help"
              isExpanded={!args.isCollapsed}
            />
          </a>
        </li>
        <li>
          <a href="www.google.com" style={{ display: "block" }}>
            <SidebarComponent.Item
              icon={
                <Avatar size="xs" bgColor="#FF8000">
                  DS
                </Avatar>
              }
              label="My profile"
              isExpanded={!args.isCollapsed}
            />
          </a>
        </li>
      </ul>
    </div>
  </SidebarComponent>
);
