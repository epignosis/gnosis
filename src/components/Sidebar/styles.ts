import { css, SerializedStyles, Theme } from "@emotion/react";
import { TypographyLevels } from "@theme/utils/typography";

export const mainNavContainer = ({ sidebar }: Theme, width: string): SerializedStyles => css`
  position: sticky;
  top: 0;
  display: flex;
  flex-direction: column;
  height: 100vh;
  background: ${sidebar.background};
  color: ${sidebar.color};
  z-index: 1001;
  max-width: ${width};

  .nav-items-wrapper {
    display: flex;
    flex-direction: column;
    flex: 1;
    margin: 1.5rem 0;
  }
`;

export const navHandleContainer = ({ sidebar }: Theme): SerializedStyles => css`
  background-color: transparent;
  padding: 0;
  border: 0;

  .nav-item {
    height: 100%;
    border-bottom: 1px solid ${sidebar.border};
  }
`;

export const navItemContainer = (
  { sidebar, typeScaleSizes }: Theme,
  fontSize: TypographyLevels,
): SerializedStyles => css`
  display: flex;
  align-items: center;
  color: ${sidebar.color};
  padding: 0.25rem 1.5rem;

  .icon-container {
    height: 2rem;
  }

  &:hover,
  &.selected,
  &.focus-visible {
    background: ${sidebar.backgroundHover};
    color: ${sidebar.colorHover};
  }

  .title {
    margin-inline-start: 0.5rem;
    font-size: ${typeScaleSizes[fontSize]};
    display: -webkit-box;
    -webkit-line-clamp: 2;
    -webkit-box-orient: vertical;
    overflow: hidden;
    text-overflow: ellipsis;
    word-break: break-word;
  }
`;
