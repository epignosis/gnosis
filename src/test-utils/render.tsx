import React, { FC, ReactElement, ComponentType } from "react";
import { render, RenderResult, RenderOptions } from "@testing-library/react";
import ThemeProvider from "../components/ThemeProvider/ThemeProvider";

const AllTheProviders: FC = ({ children }) => <ThemeProvider>{children}</ThemeProvider>;

const customRender = (ui: ReactElement, options: RenderOptions): RenderResult =>
  render(ui, { wrapper: AllTheProviders as ComponentType, ...options });

export * from "@testing-library/react";

export { customRender as render };
