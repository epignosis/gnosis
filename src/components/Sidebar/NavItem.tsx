import React, { FC, ReactNode } from "react";
import { m, AnimatePresence, Variants, useReducedMotion } from "framer-motion";
import classNames from "classnames";
import { navItemContainer } from "./styles";

export type NavItemProps = FC<{
  isExpanded: boolean;
  icon: ReactNode;
  label: string | JSX.Element;
  isActive?: boolean;
}>;

const labelVariants: Variants = {
  hidden: {
    display: "none",
    opacity: 0,
    x: -10,
  },
  visible: ({ isReducedMotion }) => ({
    display: "block",
    opacity: 1,
    x: 0,
    transition: {
      duration: isReducedMotion ? 0 : 0.1,
      delay: isReducedMotion ? 0 : 0.2,
    },
  }),
  exit: ({ isReducedMotion }) => ({
    opacity: 0,
    x: -10,
    transition: {
      duration: isReducedMotion ? 0 : 0.1,
    },
    transitionEnd: {
      display: "none",
    },
  }),
};

const NavItem: NavItemProps = ({ isExpanded, label, icon, isActive = false }) => {
  const isReducedMotion = useReducedMotion();
  const containerClassNames = classNames({
    selected: isActive,
    "nav-item": true,
  });

  return (
    <div css={navItemContainer} className={containerClassNames}>
      {icon}
      <AnimatePresence>
        {isExpanded && (
          <m.div
            className="title"
            initial="hidden"
            animate="visible"
            exit="exit"
            variants={labelVariants}
            custom={{ isReducedMotion }}
          >
            {label}
          </m.div>
        )}
      </AnimatePresence>
    </div>
  );
};

export default NavItem;
