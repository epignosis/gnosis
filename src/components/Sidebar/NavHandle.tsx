import React, { FC } from "react";
import { HamburgerSVG } from "../../icons/";
import NavItem from "./NavItem";
import { navHandleContainer } from "./styles";

export type NavHandleProps = {
  isExpanded: boolean;
  navItemLabel?: string;
  toggleMainNav: () => void;
};

const NavHandle: FC<NavHandleProps> = ({ isExpanded, toggleMainNav, navItemLabel = "" }) => {
  return (
    <button className="nav-handle" css={navHandleContainer} onClick={toggleMainNav}>
      <NavItem label={navItemLabel} isExpanded={isExpanded} icon={<HamburgerSVG height={32} />} />
    </button>
  );
};

export default NavHandle;
