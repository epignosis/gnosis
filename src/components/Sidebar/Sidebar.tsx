import React, { FC } from "react";
import {
  LazyMotion,
  m,
  useReducedMotion,
  Variants,
  domAnimation,
  HTMLMotionProps,
} from "framer-motion";
import { mainNavContainer } from "./styles";
import NavItem, { NavItemProps } from "./NavItem";
import NavHandle from "./NavHandle";

export type SidebarProps = HTMLMotionProps<"nav"> & {
  isCollapsed?: boolean;
  navHandleLabel?: string;
  onToggle?: () => void;
};

type SidebarCompoundProps = {
  Item: NavItemProps;
};

const navVariants: Variants = {
  expanded: {
    minWidth: "15rem",
    transition: {
      ease: "easeInOut",
      duration: 0.2,
    },
  },
  collapsed: {
    minWidth: "4.75rem",
    transition: {
      ease: "easeInOut",
      duration: 0.2,
      delay: 0.1,
    },
  },
  a11yExpanded: {
    minWidth: "15rem",
    transition: {
      duration: 0,
    },
  },
  a11yCollapsed: {
    minWidth: "4.75rem",
    transition: {
      duration: 0,
    },
  },
};

const Sidebar: FC<SidebarProps> & SidebarCompoundProps = ({
  isCollapsed = false,
  navHandleLabel = "Menu",
  onToggle = () => void 0,
  children,
  ...rest
}) => {
  const isReducedMotion = useReducedMotion();
  let animate: string;

  if (isReducedMotion) {
    animate = isCollapsed ? "a11yCollapsed" : "a11yExpanded";
  } else {
    animate = isCollapsed ? "collapsed" : "expanded";
  }

  return (
    <LazyMotion features={domAnimation}>
      <m.nav
        css={mainNavContainer}
        initial={false}
        animate={animate}
        variants={navVariants}
        id="sidebar"
        {...rest}
      >
        <div className="nav-items-wrapper">
          <NavHandle
            isExpanded={!isCollapsed}
            navItemLabel={navHandleLabel}
            toggleMainNav={onToggle}
          />
          {children}
        </div>
      </m.nav>
    </LazyMotion>
  );
};

Sidebar.Item = NavItem;

export default Sidebar;
