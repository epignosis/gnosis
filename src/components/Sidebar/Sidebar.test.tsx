import React from "react";
import userEvent from "@testing-library/user-event";
import { Sidebar } from "./Sidebar.stories";
import { render, screen } from "@test-utils/render";

describe("<Sidebar/>", () => {
  it("renders expanded correctly", () => {
    const mockFn = jest.fn();
    render(<Sidebar onToggle={mockFn} />);

    const handle = screen.getByRole("button");
    const items = screen.getAllByRole("link");

    expect(items).toHaveLength(9);
    expect(handle).toHaveTextContent("Menu");
    expect(items[0]).toHaveTextContent("Home");
    expect(items[1]).toHaveTextContent("My courses");
    expect(items[2]).toHaveTextContent("Catalog");
    expect(items[3]).toHaveTextContent("Calendar");
    expect(items[4]).toHaveTextContent("Messages");
    expect(items[5]).toHaveTextContent("Discussions");
    expect(items[6]).toHaveTextContent("Notifications");
    expect(items[7]).toHaveTextContent("Help");
    expect(items[8]).toHaveTextContent("My profile");
  });

  it("renders collapsed correctly", () => {
    const mockFn = jest.fn();
    render(<Sidebar isCollapsed onToggle={mockFn} />);

    const handle = screen.getByRole("button");
    const items = screen.getAllByRole("link");

    expect(items).toHaveLength(9);
    expect(handle).not.toHaveTextContent("Menu");
    expect(items[0]).not.toHaveTextContent("Home");
    expect(items[1]).not.toHaveTextContent("My courses");
    expect(items[2]).not.toHaveTextContent("Catalog");
    expect(items[3]).not.toHaveTextContent("Calendar");
    expect(items[4]).not.toHaveTextContent("Messages");
    expect(items[5]).not.toHaveTextContent("Discussions");
    expect(items[6]).not.toHaveTextContent("Notifications");
    expect(items[7]).not.toHaveTextContent("Help");
    expect(items[8]).not.toHaveTextContent("My profile");
  });

  it("fires the onToggle handler", () => {
    const mockToggle = jest.fn();
    render(<Sidebar onToggle={mockToggle} />);

    const handle = screen.getByRole("button");

    userEvent.click(handle);

    expect(mockToggle).toHaveBeenCalledTimes(1);
  });

  it("matches snapshot", () => {
    const mockFn = jest.fn();
    const { container } = render(<Sidebar onToggle={mockFn} />);

    expect(container).toMatchSnapshot();
  });
});
