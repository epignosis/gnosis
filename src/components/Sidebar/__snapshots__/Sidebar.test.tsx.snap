// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Sidebar/> matches snapshot 1`] = `
<div>
  <nav
    class="my-class css-ukfle2-mainNavContainer-mainNavContainer"
    id="my-id"
    style="min-width: 16rem;"
  >
    <div
      class="nav-items-wrapper"
    >
      <button
        class="nav-handle css-1vueq8q-navHandleContainer-navHandleContainer"
      >
        <div
          class="nav-item css-sczh30-navItemContainer-navItemContainer"
        >
          <div
            class="icon-container"
          >
            <svg-file-stub
              height="32"
            />
          </div>
          <div
            class="title"
            style="display: none; opacity: 0; transform: translateX(-10px) translateZ(0);"
          >
            Menu
          </div>
        </div>
      </button>
      <div
        style="display: flex; flex-direction: column; justify-content: space-between; flex: 1;"
      >
        <ul
          style="list-style: none; padding: 0px; margin: 0px;"
        >
          <li>
            <a
              href="www.google.com"
              style="display: block;"
            >
              <div
                class="nav-item css-sczh30-navItemContainer-navItemContainer"
              >
                <div
                  class="icon-container"
                >
                  <svg-file-stub
                    height="32"
                  />
                </div>
                <div
                  class="title"
                  style="display: none; opacity: 0; transform: translateX(-10px) translateZ(0);"
                >
                  Home
                </div>
              </div>
            </a>
          </li>
          <li>
            <a
              href="www.google.com"
              style="display: block;"
            >
              <div
                class="nav-item css-sczh30-navItemContainer-navItemContainer"
              >
                <div
                  class="icon-container"
                >
                  <svg-file-stub
                    height="32"
                  />
                </div>
                <div
                  class="title"
                  style="display: none; opacity: 0; transform: translateX(-10px) translateZ(0);"
                >
                  My courses
                </div>
              </div>
            </a>
          </li>
          <li>
            <a
              href="www.google.com"
              style="display: block;"
            >
              <div
                class="nav-item css-sczh30-navItemContainer-navItemContainer"
              >
                <div
                  class="icon-container"
                >
                  <svg-file-stub
                    height="32"
                  />
                </div>
                <div
                  class="title"
                  style="display: none; opacity: 0; transform: translateX(-10px) translateZ(0);"
                >
                  Catalog
                </div>
              </div>
            </a>
          </li>
          <li>
            <a
              href="www.google.com"
              style="display: block;"
            >
              <div
                class="nav-item css-sczh30-navItemContainer-navItemContainer"
              >
                <div
                  class="icon-container"
                >
                  <svg-file-stub
                    height="32"
                  />
                </div>
                <div
                  class="title"
                  style="display: none; opacity: 0; transform: translateX(-10px) translateZ(0);"
                >
                  Calendar
                </div>
              </div>
            </a>
          </li>
        </ul>
        <ul
          style="list-style: none; padding: 0px; margin: 0px;"
        >
          <li>
            <a
              href="www.google.com"
              style="display: block;"
            >
              <div
                class="nav-item css-sczh30-navItemContainer-navItemContainer"
              >
                <div
                  class="icon-container"
                >
                  <svg-file-stub
                    height="32"
                  />
                </div>
                <div
                  class="title"
                  style="display: none; opacity: 0; transform: translateX(-10px) translateZ(0);"
                >
                  Messages
                </div>
              </div>
            </a>
          </li>
          <li>
            <a
              href="www.google.com"
              style="display: block;"
            >
              <div
                class="nav-item css-sczh30-navItemContainer-navItemContainer"
              >
                <div
                  class="icon-container"
                >
                  <svg-file-stub
                    height="32"
                  />
                </div>
                <div
                  class="title"
                  style="display: none; opacity: 0; transform: translateX(-10px) translateZ(0);"
                >
                  Discussions
                </div>
              </div>
            </a>
          </li>
          <li>
            <a
              href="www.google.com"
              style="display: block;"
            >
              <div
                class="nav-item css-sczh30-navItemContainer-navItemContainer"
              >
                <div
                  class="icon-container"
                >
                  <svg-file-stub
                    height="32"
                  />
                </div>
                <div
                  class="title"
                  style="display: none; opacity: 0; transform: translateX(-10px) translateZ(0);"
                >
                  Show tour
                </div>
              </div>
            </a>
          </li>
          <li>
            <a
              href="www.google.com"
              style="display: block;"
            >
              <div
                class="nav-item css-sczh30-navItemContainer-navItemContainer"
              >
                <div
                  class="icon-container"
                >
                  <figure
                    class="css-1al2x53-avatar-avatar"
                  >
                    <span
                      data-testid="avatar-children-container"
                    >
                      DS
                    </span>
                  </figure>
                </div>
                <div
                  class="title"
                  style="display: none; opacity: 0; transform: translateX(-10px) translateZ(0);"
                >
                  My profile
                </div>
              </div>
            </a>
          </li>
        </ul>
      </div>
    </div>
  </nav>
</div>
`;
